File: src/prompt_utils.js
// Provides prompt template functionality for client-side use

import prompts from './prompts.js';

/**
 * Get a prompt by its path and format it with the provided variables
 *
 * @param {string} promptPath - Path to the prompt in dot notation (e.g., "landmarks.discovery")
 * @param {Object} variables - Variables to substitute in the template
 * @returns {string|null} - Formatted prompt or null if not found
 */
export function GetPrompt(promptPath, variables = {}) {
  try {
    // Navigate through the prompt structure using the path
    const pathParts = promptPath.split('.');
    let current = prompts;

    for (const part of pathParts) {
      if (!(part in current)) {
        console.error(`Prompt path not found: ${promptPath}`);
        return null;
      }
      current = current[part];
    }

    // Check if we have a template
    if (typeof current !== 'object' || !current.template) {
      console.error(`No template found at prompt path: ${promptPath}`);
      return null;
    }

    // Get the template and format it with variables
    const template = current.template;

    // Check for missing variables
    const missingVars = findMissingVariables(template, variables);
    if (missingVars.length > 0) {
      console.error(
        `Missing variables in prompt template: ${missingVars.join(', ')}`
      );
      return null;
    }

    // Format the template using JavaScript template literals
    return formatTemplate(template, variables);
  } catch (error) {
    console.error(`Error getting prompt: ${error.message}`);
    return null;
  }
}

/**
 * Find missing variables in a template
 *
 * @param {string} template - Template string with ${variable} placeholders
 * @param {Object} variables - Available variables
 * @returns {Array} - Array of missing variable names
 */
function findMissingVariables(template, variables) {
  const missingVars = [];

  // Find all {variable} patterns in the template, but only simple variable names
  const variablePattern = /\{([a-zA-Z_][a-zA-Z0-9_]*)\}/g;
  let match;

  while ((match = variablePattern.exec(template)) !== null) {
    const varName = match[1].trim();

    if (!(varName in variables)) {
      if (!missingVars.includes(varName)) {
        missingVars.push(varName);
      }
    }
  }

  return missingVars;
}

/**
 * Format a template with variables using JavaScript template literals
 *
 * @param {string} template - Template string with ${variable} placeholders
 * @param {Object} variables - Variables to substitute
 * @returns {string} - Formatted template
 */
function formatTemplate(template, variables) {
  try {
    // Simple string replacement approach to avoid scope issues
    let result = template;

    // Replace each {variable} with its value
    for (const [key, value] of Object.entries(variables)) {
      const placeholder = `{${key}}`;
      const regex = new RegExp(
        placeholder.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'),
        'g'
      );
      result = result.replace(regex, value);
    }

    return result;
  } catch (error) {
    console.error(`Error formatting template: ${error.message}`);
    throw new Error(`Template formatting failed: ${error.message}`);
  }
}

/**
 * Get a system message for AI models
 *
 * @param {string} messageType - Type of system message (e.g., "travel_agent")
 * @param {Object} variables - Variables to substitute (optional)
 * @returns {string|null} - System message or null if not found
 */
export function GetSystemMessage(messageType, variables = {}) {
  return GetPrompt(`system_messages.${messageType}`, variables);
}

/**
 * Validate that all required variables are provided for a prompt
 *
 * @param {string} promptPath - Path to the prompt
 * @param {Object} variables - Variables to check
 * @returns {Object} - Validation result with success boolean and missing variables array
 */
export function ValidatePromptVariables(promptPath, variables = {}) {
  try {
    // Navigate to the prompt
    const pathParts = promptPath.split('.');
    let current = prompts;

    for (const part of pathParts) {
      if (!(part in current)) {
        return {
          success: false,
          error: `Prompt path not found: ${promptPath}`,
          missing: [],
        };
      }
      current = current[part];
    }

    if (typeof current !== 'object' || !current.template) {
      return {
        success: false,
        error: `No template found at prompt path: ${promptPath}`,
        missing: [],
      };
    }

    // Find missing variables
    const missingVars = findMissingVariables(current.template, variables);

    return {
      success: missingVars.length === 0,
      error:
        missingVars.length > 0
          ? `Missing variables: ${missingVars.join(', ')}`
          : null,
      missing: missingVars,
    };
  } catch (error) {
    return {
      success: false,
      error: `Validation error: ${error.message}`,
      missing: [],
    };
  }
}

File: utils/prompt_utils.py
import os
import tomllib
import logging
from typing import Dict, Any, Optional

logger = logging.getLogger(__name__)

# Cache for loaded prompts
_prompts_cache = None


def load_prompts(PROMPTS_FILE="public/prompts.toml") -> Dict[str, Any]:
    """
    Load prompts from the prompts.toml file

    Returns:
        Dictionary containing all prompts and their metadata
    """
    global _prompts_cache
    from app import is_production_environment

    # Return cached prompts if available
    if is_production_environment() and _prompts_cache is not None:
        return _prompts_cache

    # Check if prompts file exists
    if not os.path.exists(PROMPTS_FILE):
        logger.error(f"Prompts file not found: {PROMPTS_FILE}")
        return {}

    try:
        # Load prompts from TOML file
        with open(PROMPTS_FILE, "rb") as f:
            _prompts_cache = tomllib.load(f)
        logger.debug(
            f"Loaded {len(_prompts_cache)} prompt categories from {PROMPTS_FILE}"
        )
        return _prompts_cache
    except Exception as e:
        logger.error(f"Error loading prompts: {str(e)}")
        return {}


def get_prompt(prompt_path: str, **variables) -> Optional[str]:
    """
    Get a prompt by its path and format it with the provided variables

    Args:
        prompt_path: Path to the prompt in dot notation (e.g., "category.subcategory.prompt_name")
        **variables: Variables to format the prompt with

    Returns:
        Formatted prompt or None if not found
    """
    prompts = load_prompts()

    # Navigate through the prompt structure using the path
    path_parts = prompt_path.split(".")
    current = prompts

    for part in path_parts:
        if part not in current:
            logger.error(f"Prompt path not found: {prompt_path}")
            return None
        current = current[part]

    # Check if we have a template
    if not isinstance(current, dict) or "template" not in current:
        logger.error(f"No template found at prompt path: {prompt_path}")
        return None

    # Get the template and format it with variables
    template = current["template"]

    try:
        return template.format(**variables)
    except KeyError as e:
        logger.error(f"Missing variable in prompt template: {str(e)}")
        return None
    except Exception as e:
        logger.error(f"Error formatting prompt: {str(e)}")
        return None


def get_system_message(message_type: str) -> Optional[str]:
    """
    Get a system message for AI models

    Args:
        message_type: Type of system message in dot notation

    Returns:
        System message or None if not found
    """
    return get_prompt(f"system_messages.{message_type}")
